blueocean-dashboard/src/main/js/util/smart-fetch.js

```diff
@@ -0,0 +1,277 @@
+        const error = new Error(response.statusText);
+        error.response = response;
+        throw error;
+    }
+    return response;
+};
+
+/**
+ * Get the JSON response body
+ */
+const parseJSON = (rsp) => {
+    try {
+        return rsp.json();
+    } catch (err) {
+        console.error('Unable to parse JSON: ', rsp.body, err); // eslint-disable-line no-console
+        throw new Error('Invalid JSON payload', err);
+    }
+};
+
+/**
+ * Static fetch options used for every request
+ */
+const fetchOptions = { credentials: 'same-origin' };
+
+/**
+ * DuplicateCallTracker maintains active fetches against a particular key
+ */
+export class DuplicateCallTracker {
+    constructor() {
+        /**
+         * Onload callbacks cache. Used to ensure we don't
+         * issue multiple in-parallel requests for the same
+         * class metadata.
+         */
+        this.fetchCallbacks = {};
+        this.promises = {};
+    }
+
+    /**
+     * Generalization of consolidation of duplicate requests
+     */
+    execute(key, promiseProvider, onComplete) {
+        let callbacks = this.fetchCallbacks[key];
+        if (!callbacks) {
+            // This is the first request for this type. Initialize the
+            // callback cache and then issue the request
+            callbacks = this.fetchCallbacks[key] = onComplete ? [onComplete] : [];
+
+            promiseProvider()
+                .then((data) => {
+                    // Notify all callbacks
+                    for (let i = 0; i < callbacks.length; i++) {
+                        try {
+                            callbacks[i](data);
+                        } catch (e) {
+                            console.error('Unexpected Error fulfilling callbacks', e); // eslint-disable-line no-console
+                        }
+                    }
+                    delete this.fetchCallbacks[key];
+                    return data; // for downstream .then
+                });
+        } else {
+            // We already have an in-flight request, just store the onComplete callback
+            callbacks.push(onComplete);
+        }
+    }
+
+
+    /**
+     * Generalization of consolidation of duplicate requests
+     */
+    promise(key, promiseProvider, onComplete) {
+        const holder = this.promises[key] || (this.promises[key] = { count: 0, promise: promiseProvider() }); // get or create
+        holder.count++;
+        holder.promise = holder.promise.then((data) => {
+            holder.count--;
+            if (holder.count === 0) {
+                delete this.promises[key];
+            }
+            try {
+                onComplete(data);
+            } catch (e) {
+                // TODO how best to handle failures
+                console.error(e); // eslint-disable-line no-console
+            }
+            return data; // for downstream .then
+        });
+        return holder.promise;
+    }
```

Comment:
So it uses one promise instance to handle all the callers, stringing them together on the `then`? If so, that seems baaaadddddd to me.

Would also explain why there's a TODO wrt handling the errors i.e. if each caller was given their own `Promise` then passing back the errors would be trivial.